### DISCUSSION TITLE
Bring your own branded / opaque types?

### DISCUSSION THREAD

**Discussion by jrnail23** [colinhacks/zod#2002] [Q&A] [answered]
Hi, I'm just digging into Zod for the first time, and I see you support branded types (yay!).
My project already has a bunch of branded types defined using [`ts-essentials` opaque types](https://www.npmjs.com/package/ts-essentials#Opaque-types), with parsing functions for those types via the "companion object" pattern.

My question is, is there a way to have Zod use my already existing branded/opaque types and their accompanying parsers?

**Comment by igalklebanov** [comment, marked as answer]
Hey 👋 

Have you looked at https://github.com/colinhacks/zod#custom-schemas ?

```ts
const schema = z.object({
  id: z.custom<brandedIDType>(brandedIdType),
  name: z.string().min(1),
  dob: z.coerce.date(),
})
```

**Reply by JacobWeisenburger** [reply]
@igalklebanov Great job!

**Comment by JacobWeisenburger** [comment]
this seems like more of a QA discussion, is it ok if I move it there?
